src/Features/CSharp/Portable/OrderModifiers/OrderModifiersHelper.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.OrderModifiers
+{
+    internal static class OrderModifiersHelper
+    {
+        private static readonly char[] s_comma = { ',' };
+        private static readonly object s_gate = new object();
+
+        private static string s_lastValue;
+        private static Dictionary<int, int> s_lastPreferredOrder;
+
+        public static bool TryGetOrComputePreferredOrder(string value, out Dictionary<int, int> preferredOrder)
+        {
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                preferredOrder = null;
+                return false;
+            }
+
+            lock (s_gate)
```

Comment:
❗️ This is called on a highly-concurrent code path during analysis. Can't have a lock on the fast path.